### PR TITLE
feat: Add dep group support

### PR DESCRIPTION
**Pull request by TheWaWaR**
- [x] Add dep group support
- [x] Add tx pool tests for dep group

### CODE REVIEW

File: core/src/cell.rs [hunk 1]
```diff
@@ -341,11 +342,58 @@ impl HeaderChecker for BlockHeadersChecker {
...
+    }
+
+    let dep_len = data.len() / OUT_POINT_LEN;
+    let mut out_points = Vec::with_capacity(dep_len);
+    for dep_idx in 0..dep_len {
+        let item_start = dep_idx * OUT_POINT_LEN;
+        let item_end = (dep_idx + 1) * OUT_POINT_LEN;
+        out_points.push(
+            OutPoint::from_group_data(&data[item_start..item_end])
+                .expect("parse group item data failed"),
```
Comment (u2) [inline core/src/cell.rs]: should return Err even if we know the length is correct.

Comment (TheWaWaR) [reply core/src/cell.rs]: We already checked the length above, so it can not be error.

Comment (u2) [reply core/src/cell.rs]: They have only a fine distinction between them. If we refactor the from_group_data, it's very easily overlooked here, this `expect` depends on the implementation of the `from_group_data`.It's not serious.

File: shared/src/chain_state.rs [hunk 2]
```diff
@@ -584,8 +595,29 @@ impl ChainState {
 
         let retain: Vec<Transaction> = detached.difference(&attached).cloned().collect();
 
+        let txs_iter = attached.iter().map(|tx| {
+            let related_out_points = tx.cell_deps_iter().fold(
```
Comment (jjyr) [inline shared/src/chain_state.rs]: Provide a `resolve_transacrtion` like function to resolve dep_group of a tx. for example `fn resolve_tx_dep_group(&tx: Transaction, &store: Store) -> Vec<OutPoints>`

File: core/src/cell.rs [hunk 3]
```diff
@@ -341,11 +342,58 @@ impl HeaderChecker for BlockHeadersChecker {
...
+    /// Invalid dep group cell data length
+    InvalidDepGroup(OutPoint),
     Dead(OutPoint),
     Unknown(Vec<OutPoint>),
     OutOfOrder(OutPoint),
 }
 
+pub fn parse_dep_group_data(data: &[u8]) -> Result<Vec<OutPoint>, usize> {
+    // tx hash (32 bytes) + output index (4 bytes)
+    const OUT_POINT_LEN: usize = mem::size_of::<H256>() + mem::size_of::<u32>();
```
[the remark this answers is not in the document]
Comment (TheWaWaR) [reply core/src/cell.rs]: size_of for primitive type is not a problem.

File: core/src/cell.rs [hunk 4]
```diff
@@ -341,11 +342,58 @@ impl HeaderChecker for BlockHeadersChecker {
...
+    Ok(out_points)
+}
+
+fn resolve_dep_group<
+    F: FnMut(&OutPoint, bool) -> Result<Option<Box<CellMeta>>, UnresolvableError>,
+>(
+    out_point: &OutPoint,
+    mut cell_resolver: F,
+) -> Result<Vec<CellMeta>, UnresolvableError> {
+    let data = match cell_resolver(out_point, true)? {
```
[the remark this answers is not in the document]
Comment (TheWaWaR) [reply core/src/cell.rs]: It's not just load the data, if the cell is dead, we must return error.

File: core/src/transaction.rs [hunk 5]
```diff
@@ -98,6 +98,22 @@ impl OutPoint {
     pub fn cell_key(&self) -> CellKey {
         CellKey::calculate(&self.tx_hash, self.index)
     }
+
+    /// Convert from dep group data
+    pub fn from_group_data(data: &[u8]) -> Result<OutPoint, usize> {
+        if data.len() != mem::size_of::<H256>() + mem::size_of::<u32>() {
```
[the remark this answers is not in the document]
Comment (TheWaWaR) [reply core/src/transaction.rs]: Thought about it, I'll just put it in `cell.rs`

File: shared/src/chain_state.rs [hunk 6]
```diff
@@ -518,7 +529,7 @@ impl ChainState {
             },
             Err(err) => {
                 match &err {
-                    UnresolvableError::Dead(_) => {
+                    UnresolvableError::Dead(_) | UnresolvableError::InvalidDepGroup(_) => {
```
[the remark this answers is not in the document]
Comment (TheWaWaR) [reply shared/src/chain_state.rs]: Sounds right. 

(There was a reason to put it here, I should put a comment back then)

File: core/src/cell.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jjyr) [reply core/src/cell.rs]: I see, but we can still return a `Result<Bytes, UnresolvableError>` to represent that, my concern is the bool argument and `mem_data` is hard to understand, it's more readable to hide these into a function.

File: core/src/cell.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jjyr) [reply core/src/cell.rs]: But `H256` is not a primitive type of rust, and we can't guarantee that @yangby-cryptape does not change the inner data structure in the future. 

Even so, here we check the binary format, from `semantics` the concept it's not equal to mem size, use a const `36` is more straightforward.

File: core/src/cell.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (TheWaWaR) [reply core/src/cell.rs]: Below this code in the same function, we also want resolve cell without data.

The bool argument will changed to an enumerate in the near future.

File: core/src/cell.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (TheWaWaR) [reply core/src/cell.rs]: Changed to just 36

**Pull request merged**